### PR TITLE
Implements `excludes`

### PR DESCRIPTION
**Pull request by calve**
As suggested in #132.

### CODE REVIEW

**Comment by nicolaiarocci** [comment]
Excellent.

File: cerberus/cerberus.py [hunk 1]
```diff
@@ -758,6 +758,17 @@ def _validate_valueschema(self, schema, field, value):
                 if len(validator.errors):
                     self._error(field, validator.errors)
 
+    def _validate_excludes(self, exclude, field, value):
+        targets = []
+        if isinstance(exclude, list):
```
Comment (funkyfuture) [inline cerberus/cerberus.py]: please test for `isinstance(exclude, Sequence) and not isinstance(exclude, _str_types)`

File: cerberus/cerberus.py [hunk 2]
```diff
@@ -758,6 +758,17 @@ def _validate_valueschema(self, schema, field, value):
                 if len(validator.errors):
                     self._error(field, validator.errors)
 
+    def _validate_excludes(self, exclude, field, value):
```
Comment (funkyfuture) [inline cerberus/cerberus.py]: should be named `excludes` as a list is expected.

File: cerberus/cerberus.py [hunk 3]
```diff
@@ -758,6 +758,17 @@ def _validate_valueschema(self, schema, field, value):
...
+        targets = []
```
Comment (funkyfuture) [inline cerberus/cerberus.py]: it's enough to wrap a single string into a list with the name `excludes`, no need for an extra object `target` imo.

**Comment by funkyfuture** [comment]
please also add a check to `SchemaError.validate` for the `excludes`-rule. please look at #135 for the scheme i used to refactor schema-errors.

otherwise, lgtm. and i'm even baffled that `required` is dismissed correctly without extra effort. 8-)

File: docs/usage.rst [hunk 4]
```diff
@@ -616,6 +616,62 @@ Thus you can use this to validate a document against several schemas without imp
...
+
+.. doctest::
+
+    >>> v = Validator()
+    >>> schema = {'this_field': {'type': 'dict',
+    ...                          'excludes': 'that_field'},
+    ...           'that_field': {'type': 'dict',
+    ...                          'excludes': 'this_field'}}
+    >>> v.validate({'this_field': {}, 'that_field': {}}, schema)
+    ... False
```
Comment (funkyfuture) [inline docs/usage.rst]: that has to be

```
False
```

also true for all following output of returned objects. with leading dots it's interpreted as statement, not as expected result.

**Comment by calve** [comment]
Thank you very much for the review. I considered your observations and pushed the appropriate changes.

However, I am not quite sure what you mean by 

> add a check to ``SchemaError.validate``

I added a test for the error string. Please elaborate if this was not what you meant.

**Comment by funkyfuture** [comment]
ah, sorry. i meant [`DefinitionSchema.validate`](https://github.com/nicolaiarocci/cerberus/pull/137/files#diff-aebc53cd4926f3a579adb7ba188de369R809); ensure that the defintion for an `excludes`-constraint is either a Hashable or a (non-string-)sequence of `Hashable`s to avoid runtime-errors on validation.

**Comment by funkyfuture** [comment]
though i don't see a use-case for it, you might add handling of the rule for unknown fields for the sake of completeness. that would let *any* arbritary field to exclude one or more defined fields.

File: cerberus/cerberus.py [hunk 5]
```diff
@@ -758,6 +758,16 @@ def _validate_valueschema(self, schema, field, value):
                 if len(validator.errors):
                     self._error(field, validator.errors)
 
+    def _validate_excludes(self, excludes, field, value):
+        if isinstance(excludes, _str_type):
```
Comment (funkyfuture) [inline cerberus/cerberus.py]: should test for `Hashable` as any hashable object can be a mapping's key.

**Comment by nicolaiarocci** [comment]
Could you please rebase this one of top of current `master`? Thanks!

**Comment by calve** [comment]
Rebased.

Running ``tox`` test suite, those two errors are raised;

```
Document: usage
---------------
**********************************************************************
File "usage.rst", line 654, in default
Failed example:
    v.validate({'this_field': {}}, schema)
Expected:
    True
Got:
    False
**********************************************************************
File "usage.rst", line 656, in default
Failed example:
    v.validate({'that_field': {}}, schema)
Expected:
    True
Got:
    False
**********************************************************************
```

which are not consistent with the results expected by the test suite. Any idea how to fix those ?

**Comment by funkyfuture** [comment]
in such cases  i adapt the doctests as unittests to investigate.

**Comment by nicolaiarocci** [comment]
hmm I just ran tox and didn't get any issues (travis agrees too).

**Comment by nicolaiarocci** [comment]
For line 654:

    >>> v.errors
    {'that_field': 'required field'}

**Comment by nicolaiarocci** [comment]
The similar test succeeds because `assertSuccess` passes `update=True` to the `validate` method. If `update` is `False`, then it will fail (as it happens to be in the doctest).

**Comment by calve** [comment]
It looks like when both fields are required, the validatiion always fails. That makes sence, but the following test case should fails.

    def test_required_excludes(self):
        schema = {'this_field': {'type': 'dict',
                                 'excludes': 'that_field',
                                 'required': True},
                  'that_field': {'type': 'dict',
                                 'excludes': 'this_field',
                                 'required': True}}
        document1 = {'this_field': {}}
        document2 = {'that_field': {}}
        document3 = {'that_field': {}, 'this_field': {}}
        self.assertSuccess(document1, schema)
        self.assertSuccess(document2, schema)
        self.assertFail({}, schema)
        self.assertFail(document3, schema)

in particular the following lines : 
    
        self.assertSuccess(document1, schema)
        self.assertSuccess(document2, schema)

Actually, the test suite still succeed if I changed those ``assertSuccess`` for ``assertFail``

**Comment by nicolaiarocci** [comment]
Yup see my comment on `assertSuccess` above

**Comment by calve** [comment]
This makes sense. Do you have a suggestion about it ?

I am about to patch the ``assertSuccess`` method to accept an ``update`` parameter.

**Comment by nicolaiarocci** [comment]
Patching `assertSuccess` is probably a good thing. You will also want to edit/change the usage.rst example.

**Comment by calve** [comment]
Please have a look at the latest commit : running ``assertSuccess`` with ``update=False`` raised errors concerning mutually excluded and required fields (in the ``test_required_excludes`` function). 

``required`` fields were not properly dismissed, which should be fixed now.

**Comment by calve** [comment]
Thanks for the review once again. I applied your suggestions.

File: cerberus/cerberus.py [hunk 6]
```diff
@@ -818,6 +830,28 @@ def _validate_valueschema(self, schema, field, value):
                 if len(validator.errors):
                     self._error(field, validator.errors)
 
+    def _validate_excludes(self, excludes, field, value):
```
Comment (funkyfuture) [inline cerberus/cerberus.py]: one thing i missed, within their functional grouped sections methods are sorted alphabetically. thus, this should be moved to line 578.

Comment (calve) [inline cerberus/cerberus.py]: Updated

**Comment by nicolaiarocci** [comment]
Thanks!

**Comment by nicolaiarocci** [comment]
PS: feel free to PR to replace your nickname if so you wish (could not get a hold of your name from the logs). Or I can edit it for you :)

**Comment by funkyfuture** [comment]
great, thank you!

**Comment by calve** [comment]
Thanks for all the reviews and the approval.

**Pull request merged**